flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/StopWithSavepointContextTest.java

```diff
@@ -0,0 +1,249 @@
+                            Collections.singletonList(ExecutionState.RUNNING));
+                    testInstance.handleSavepointCreation("savepoint-path", null);
+                });
+    }
+
+    private void assertNoTerminationHandling(Runnable stopWithSavepointCompletion)
+            throws Exception {
+        final ManuallyTriggeredScheduledExecutor restartExecutor =
+                new ManuallyTriggeredScheduledExecutor();
+        scheduler =
+                SchedulerTestingUtils.createSchedulerBuilder(
+                                jobGraph, ComponentMainThreadExecutorServiceAdapter.forMainThread())
+                        // we're expecting a global fail-over and,
+                        // therefore, have to enable restarting
+                        .setRestartBackoffTimeStrategy(new TestRestartBackoffTimeStrategy(true, 0))
+                        .setDelayExecutor(restartExecutor)
+                        .setFutureExecutor(new DirectScheduledExecutorService())
+                        .build();
+        scheduler.startScheduling();
+
+        testInstance = new StopWithSavepointContext(jobGraph.getJobID(), scheduler, log);
+
+        disableCheckpointScheduler();
+
+        stopWithSavepointCompletion.run();
+
+        // the task gets cancelled before triggering the restart
+        ExecutionAttemptID executionAttemptID =
+                scheduler
+                        .getExecutionGraph()
+                        .getAllExecutionVertices()
+                        .iterator()
+                        .next()
+                        .getCurrentExecutionAttempt()
+                        .getAttemptId();
+        scheduler.updateTaskExecutionState(
+                new TaskExecutionState(
+                        jobGraph.getJobID(), executionAttemptID, ExecutionState.CANCELED));
+
+        restartExecutor.triggerScheduledTasks();
+
+        try {
+            testInstance.getResult().get();
+            fail("An ExecutionException is expected.");
+        } catch (Throwable e) {
+            final Optional<FlinkException> expectedFlinkException =
+                    ExceptionUtils.findThrowable(e, FlinkException.class);
+            final String expectedMessage =
+                    String.format(
+                            "Inconsistent execution state after stopping with savepoint. At least one execution is still in one of the following states: RUNNING. A global fail-over is triggered to recover the job %s.",
+                            jobGraph.getJobID());
+            assertTrue(expectedFlinkException.isPresent());
+            assertThat(
+                    expectedFlinkException.get(), FlinkMatchers.containsMessage(expectedMessage));
+        }
+
+        // the global fail-over puts all tasks into DEPLOYING state again
+        assertExecutionStates(scheduler, ExecutionState.DEPLOYING);
+
+        // the CheckpointCoordinator should be enabled again
+        assertCheckpointSchedulingEnabled();
```

Comment:
This looks quite complicated for a simple unit test. This is usually a sign that the subject under test is not yet super easy to test.

Reply:
I refactored the code accordingly introducing `CheckpointScheduling` as an interface and switching to `SchedulerNG` for the global failover triggering. I hesitated to move `SchedulerNG.handleGlobalFailure` as it felt to be at the right place in `SchedulerNG` already.